File: FoosballGames/FoosballGame.cs
using FoosballGames.Contracts;
using FoosballGames.Contracts.Exceptions;
using NodaTime;

namespace FoosballGames;

public record FoosballGame(Guid Id, LocalDateTime Start, LocalDateTime? End, ISets Sets)
{
    public static FoosballGame Initialize(CreateFoosballGame command)
    {
        return new(command.Id, LocalDateTime.FromDateTime(command.Start), null,
            new FirstSetRunning(RunningSet.New()));
    }

    public FoosballGame AddPointForTeam(Team team)
    {
        var sets = Sets.AddPointForTeam(team);
        var end = sets is FinishedSets ? LocalDateTime.FromDateTime(DateTime.Now) : (LocalDateTime?) null;
        return this with {Sets = sets, End = end};
    }
}

public interface ISets
{
    public ISets AddPointForTeam(Team team);
}

public record FirstSetRunning(RunningSet Set) : ISets
{
    public ISets AddPointForTeam(Team team)
    {
        var currentSet = Set.AddPointForTeam(team);

        return currentSet switch
        {
            FinishedSet finishedSet => new SecondSetRunning(finishedSet, RunningSet.New()),
            RunningSet runningSet => this with {Set = runningSet},
            _ => throw new ArgumentOutOfRangeException(nameof(currentSet), currentSet, "Unexpected set type.")
        };
    }
}

public record SecondSetRunning(FinishedSet FirstSet, RunningSet SecondSet) : ISets
{
    public ISets AddPointForTeam(Team team)
    {
        var currentSet = SecondSet.AddPointForTeam(team);

        return currentSet switch
        {
            FinishedSet finishedSet when finishedSet.WinnerTeam != FirstSet.WinnerTeam =>
                new ThirdSetRunning(FirstSet, finishedSet, RunningSet.New()),
            FinishedSet finishedSet when finishedSet.WinnerTeam == FirstSet.WinnerTeam =>
                new FinishedSets(FirstSet, finishedSet, null, finishedSet.WinnerTeam),
            RunningSet runningSet => this with {SecondSet = runningSet},
            _ => throw new ArgumentOutOfRangeException(nameof(currentSet), currentSet, "Unexpected set type.")
        };
    }
}

public record ThirdSetRunning(FinishedSet FirstSet, FinishedSet SecondSet, RunningSet ThirdSet) : ISets
{
    public ISets AddPointForTeam(Team team)
    {
        var currentSet = ThirdSet.AddPointForTeam(team);

        return currentSet switch
        {
            FinishedSet finishedSet => new FinishedSets(FirstSet, SecondSet, finishedSet, finishedSet.WinnerTeam),
            RunningSet runningSet => this with {ThirdSet = runningSet},
            _ => throw new ArgumentOutOfRangeException(nameof(currentSet), currentSet, "Unexpected set type.")
        };
    }
}

public record FinishedSets(FinishedSet FirstSet, FinishedSet SecondSet, FinishedSet? ThirdSet,
    Team WinnerTeam) : ISets
{
    public ISets AddPointForTeam(Team team) => throw new CannotAddPointToFinishedSet();
}

public interface ISet
{
}

public record RunningSet(byte RedTeamScore, byte BlueTeamScore) : ISet
{
    private const byte MaxPointsScore = 10;
    public static RunningSet New() => new(0, 0);

    public ISet AddPointForTeam(Team team)
    {
        var (blueScore, redScore) = team switch
        {
            Team.Blue => ((byte)(BlueTeamScore + 1), RedTeamScore),
            Team.Red => (BlueTeamScore, (byte)(RedTeamScore + 1)),
            _ => throw new ArgumentOutOfRangeException(nameof(team), team, "Unexpected team.")
        };

        if (blueScore == MaxPointsScore || redScore == MaxPointsScore)
            return new FinishedSet(RedTeamScore: redScore, BlueTeamScore: blueScore, team);

        return new RunningSet(RedTeamScore: redScore, BlueTeamScore: blueScore);
    }
}

public record FinishedSet(byte RedTeamScore, byte BlueTeamScore, Team WinnerTeam) : ISet;

File: FoosballGames/FoosballGame.fs
namespace FoosballGames

open System
open FoosballGames.Contracts
open FoosballGames.Contracts.Exceptions
open NodaTime

type ISet =
    interface end

type FinishedSet =
    {RedTeamScore: byte; BlueTeamScore: byte; WinnerTeam: Team}
    interface ISet

type RunningSet =
    {RedTeamScore: byte; BlueTeamScore: byte}
    interface ISet
    static member New () = {RedTeamScore = 0uy; BlueTeamScore = 0uy}
    static member MaxPointsScore = 10uy
    member this.AddPointForTeam team : ISet =
        let blueScore, redScore =
            match team with
            | Team.Blue -> (this.BlueTeamScore + 1uy, this.RedTeamScore)
            | Team.Red -> (this.BlueTeamScore, this.RedTeamScore + 1uy)
            | _ -> raise <| ArgumentOutOfRangeException(nameof team, team, "Unexpected team.")
        
        match (blueScore = RunningSet.MaxPointsScore, redScore = RunningSet.MaxPointsScore) with
        | true, _ | _, true ->
            {RedTeamScore = redScore; BlueTeamScore = blueScore; WinnerTeam = team}
        | _ -> {RedTeamScore = redScore; BlueTeamScore = blueScore} :> ISet

type ISets =
    abstract member AddPointForTeam: team: Team -> ISets

type FinishedSets =
    {FirstSet: FinishedSet; SecondSet: FinishedSet; ThirdSet: FinishedSet option; WinnerTeam: Team}
    
    interface ISets with
        member this.AddPointForTeam _ : ISets =
            raise <| CannotAddPointToFinishedSet()

type ThirdSetRunning =
    {FirstSet: FinishedSet; SecondSet: FinishedSet; ThirdSet: RunningSet}
    
    interface ISets with
        member this.AddPointForTeam team =
            let currentSet = this.ThirdSet.AddPointForTeam team
            match currentSet with
            | :? FinishedSet as finished ->
                { FirstSet = this.FirstSet; SecondSet = this.SecondSet
                  ThirdSet = Some(finished); WinnerTeam = finished.WinnerTeam }
            | :? RunningSet as running -> { this with ThirdSet = running }
            | _ -> raise <| ArgumentOutOfRangeException(nameof currentSet, currentSet, "Unexpected set type.")    

type SecondSetRunning =
    {FirstSet: FinishedSet; SecondSet: RunningSet}
    interface ISets with
        member this.AddPointForTeam team =
            let currentSet = this.SecondSet.AddPointForTeam team
            match currentSet with
            | :? FinishedSet as finished when finished.WinnerTeam <> this.FirstSet.WinnerTeam ->
                { FirstSet = this.FirstSet; SecondSet = finished; ThirdSet = RunningSet.New() }
            | :? FinishedSet as finished when finished.WinnerTeam = this.FirstSet.WinnerTeam ->
                { FirstSet = this.FirstSet; SecondSet = finished; ThirdSet = None; WinnerTeam = finished.WinnerTeam }
            | :? RunningSet as running -> { this with SecondSet = running }
            | _ -> raise <| ArgumentOutOfRangeException(nameof currentSet, currentSet, "Unexpected set type.")
            
type FirstSetRunning =
    {Set: RunningSet}
    
    interface ISets with
        member this.AddPointForTeam(team) =
            let currentSet = this.Set.AddPointForTeam team
            match currentSet with
            | :? FinishedSet as finished -> { FirstSet = finished; SecondSet = RunningSet.New() }
            | :? RunningSet as running -> { this with Set = running }
            | _ -> raise <| ArgumentOutOfRangeException(nameof currentSet, currentSet, "Unexpected set type.")

type FoosballGame =
    {Id: Guid; Start: LocalDateTime; End: LocalDateTime option; Sets: ISets}
    
    static member Initialize (command: CreateFoosballGame) =
        {Id = command.Id
         Start = LocalDateTime.FromDateTime(command.Start)
         End = None
         Sets = { Set = RunningSet.New() }}
        
    member this.AddPointForTeam team =
        let sets = this.Sets.AddPointForTeam team
        let endTime = if sets :? FinishedSets then Some(LocalDateTime.FromDateTime(DateTime.Now)) else None
        { this with Sets = sets; End = endTime }

(*
module FoosballGames.FoosballGame

open System
open FoosballGames.Contracts
open NodaTime

type RunningSet = {BlueTeamScore: byte; RedTeamScore: byte}
type FinishedSet = {BlueTeamScore: byte; RedTeamScore: byte; Winner: Team}

type Set =
    | Running of RunningSet
    | Finished of FinishedSet

type Game =
    | FirstSet
    | SecondSet
    | ThirdSet
    | Finished

type FoosballGame = { Id: Guid; Start: LocalDateTime; End: LocalDateTime; }
*)